File: script/renderer/animation_clip_2d.lua
require("lua_extension")
require("cpp_class")
require("renderer/sprite")

--- @class AnimationClip2D 2D动画片段
AnimationClip2D = class("AnimationClip2D", CppClass)

function AnimationClip2D:ctor(...)
    AnimationClip2D.super.ctor(self, ...)
    if self.cpp_class_instance_ == nil then
        self.cpp_class_instance_ = Cpp.AnimationClip2D()
    end
end

function AnimationClip2D:AddFrame(sprite, duration)
    self.cpp_class_instance_:AddFrame(sprite:cpp_class_instance(), duration)
end

function AnimationClip2D:set_is_looping(loop)
    self.cpp_class_instance_:set_is_looping(loop)
end

function AnimationClip2D:is_looping()
    return self.cpp_class_instance_:is_looping()
end

function AnimationClip2D:set_name(name)
    self.cpp_class_instance_:set_name(name)
end

function AnimationClip2D:name()
    return self.cpp_class_instance_:name()
end

File: src/renderer/animation_clip_2d.cpp
#include "animation_clip_2d.h"
#include "sprite.h"
#include <rttr/registration>

using namespace rttr;

RTTR_REGISTRATION
{
    registration::class_<AnimationClip2D>("AnimationClip2D")
        .constructor<>()(rttr::policy::ctor::as_raw_ptr)
        .property("is_looping", &AnimationClip2D::is_looping, &AnimationClip2D::set_is_looping)
        .property("name", &AnimationClip2D::name, &AnimationClip2D::set_name);
}

AnimationClip2D::AnimationClip2D() {
}

AnimationClip2D::~AnimationClip2D() {
}

void AnimationClip2D::AddFrame(Sprite* sprite, float duration) {
    frames_.push_back({sprite, duration});
}
